source/delimiter-case.d.ts

```diff
@@ -0,0 +1,64 @@
+export type UpperCaseChars = 'A' | 'B' | 'C' | 'D' | 'E' | 'F' | 'G' | 'H' | 'I' | 'J' | 'K' | 'L' | 'M' | 'N' | 'O' | 'P' | 'Q' | 'R' | 'S' | 'T' | 'U' | 'V' | 'X' | 'Y' | 'Z';
+export type WordSeparators = '-' | '_' | ' ';
+
+export type SplitIncludingDelimiters<Source extends string, Delimiter extends string> =
+	Source extends '' ? [] :
+	Source extends `${infer FirstPart}${Delimiter}${infer SecondPart}` ?
+	(
+		Source extends `${FirstPart}${infer UsedDelimiter}${SecondPart}`
+			? UsedDelimiter extends Delimiter
+				? Source extends `${infer FirstPart}${UsedDelimiter}${infer SecondPart}`
+					? [...SplitIncludingDelimiters<FirstPart, Delimiter>, UsedDelimiter, ...SplitIncludingDelimiters<SecondPart, Delimiter>]
+					: never
+				: never
+			: never
+	) :
+	[Source];
+
+type StringPartToDelimiterCase<StringPart extends string, UsedWordSeparators extends string, UsedUpperCaseChars extends string, Delimiter extends string> =
+	StringPart extends UsedWordSeparators ? Delimiter :
+	StringPart extends UsedUpperCaseChars ? `${Delimiter}${Lowercase<StringPart>}` :
+	StringPart;
+
+type StringArrayToDelimiterCase<Parts extends any[], UsedWordSeparators extends string, UsedUpperCaseChars extends string, Delimiter extends string> =
+	Parts extends [`${infer FirstPart}`, ...infer RemainingParts]
+		? `${StringPartToDelimiterCase<FirstPart, UsedWordSeparators, UsedUpperCaseChars, Delimiter>}${StringArrayToDelimiterCase<RemainingParts, UsedWordSeparators, UsedUpperCaseChars, Delimiter>}`
+		: '';
+
+/**
+Converts a string literal from any typical casing to one based on a custom delimiter
+
+This can be useful when eg. converting a camel cased object property to an oddly cased one.
+
+Also see KebabCase and SnakeCase.
```

Comment:
Could we use `@see` here? https://devblogs.microsoft.com/typescript/announcing-typescript-4-1/#jsdoc-see-tag

Reply:
@sindresorhus I'm a bit unsure how to do that actually, in the reference they have imported the module that they are referring to, but that's not the case here.

I'll just add this for now:

```
@see KebabCase
@see SnakeCase
```